### PR TITLE
TF: Unpack model inputs through a decorator 

### LINKED ISSUE #15908: RFC -- TF: unpack model inputs through a decorator
# RFC

This is an RFC concerning how model/layer inputs are handled in TF. The details are below and it includes a working demo with the proposed solution -- comments (and concerns) are deeply appreciated.

cc @Rocketknight1 @LysandreJik @sgugger 

## Motivation

In all TF models and main layers, in their `call` function, we start by processing their inputs with `input_processing()` ([example](https://github.com/huggingface/transformers/blob/master/src/transformers/models/bert/modeling_tf_bert.py#L741)). This function is needed to handle multiple input formats, from a dictionary of tensors packed in the first argument of `call` to variables of diverse types passed through keyword arguments (or a combination of both), as well as to handle some control logic. 

The `input_processing()` function outputs a dictionary containing the processed function arguments, often held in the `inputs` variable. Sadly, it forces us to access all function arguments through the `input` dictionary throughout its corpus, which results in verbose and less clear code ([example](https://github.com/huggingface/transformers/blob/master/src/transformers/models/bert/modeling_tf_bert.py#L815)). Furthermore, it is not beginner-friendly -- it's not obvious at first how, why, and when we need to use the `input` dictionary, as opposed to the function arguments.

## Proposed solution

The `inputs` dictionary is essentially a dictionary containing the function arguments and keyword arguments (plus potentially extra `**kwargs` that were unpacked in the process). If we were to call `call` again using these processed inputs, no input processing should be needed. The proposed solution is precisely that -- a decorator that runs `input_processing()` with whatever is passed to the function, sending its output to `call`.

As a proof of concept, I've built the decorator and applied it to a single class in [this draft PR](https://github.com/huggingface/transformers/pull/15907) -- all associated tests pass. As you can see, the resulting code is shorter, clearer, and makes use of the expected function arguments throughout its corpus.

## Proposed plan for adoption

If we agree to go forward with this solution, here's my proposed plan:
1. Open a PR with the decorator, including its application on an NLP model and some other non-NLP model. This would confirm that the decorator works fine for multiple modalities;
2. Applying the decorator is straightforward, so we could open an issue with the `Good First Issue` label. I'd keep track of collaborators, and give it a deadline (e.g. 1 month) after which I'd take over remaining models;
3. After all models are updated, move `input_processing()` so as to be an internal function of the decorator.

### PR DESCRIPTION
**Pull request by gante**
# What does this PR do?

Unpacks TF model inputs through a decorator, improving code clarity. To be read with issue https://github.com/huggingface/transformers/issues/15908, which holds the description of the problem, the proposed solution, and future plan.

Closes https://github.com/huggingface/transformers/issues/15908

How to review: start with `modeling_tf_utils.py`, and then check the changes in the models. I've run `RUN_SLOW=1 py.test -vv tests/model_name/test_modeling_tf_model_name.py` for all involved models, and the changes were applied to BERT (plus its copy-linked architectures) and Speech2Text, showing that it works for multiple modalities.

### CODE REVIEW

**Comment by gante** [comment]
As suggested by @sgugger in #15908, I've opened the PR :)

File: src/transformers/models/speech_to_text/modeling_tf_speech_to_text.py [hunk 1]
```diff
@@ -822,81 +823,64 @@ def call(
...
-            attention_mask=attention_mask,
-            head_mask=head_mask,
-            output_attentions=output_attentions,
-            output_hidden_states=output_hidden_states,
-            return_dict=return_dict,
-            training=training,
-            kwargs_call=kwargs,
-        )
-        if "input_ids" in inputs:
-            inputs["input_features"] = inputs.pop("input_ids")
```
Comment (gante) [inline src/transformers/models/speech_to_text/modeling_tf_speech_to_text.py:838]: No more need for this ugly work-around 😎

**Review by sgugger — COMMENTED** [review]
So nice and elegant! Love how it makes the modeling code so much more readable!

**Comment by LysandreJik** [comment]
This looks great! Also pinging @patrickvonplaten for review as it's quite central

**Comment by LysandreJik** [comment]
Thanks for running the tests on it. It is way more readable like that indeed.

File: src/transformers/modeling_tf_utils.py [hunk 2]
```diff
@@ -344,6 +344,46 @@ def booleans_processing(config, **kwargs):
...
+
+    # Keras enforces the first layer argument to be passed, and checks it through `inspect.getfullargspec()`. This
+    # function does not follow wrapper chains (i.e. ignores `functools.wraps()`), meaning that without the line below
+    # Keras would attempt to check the first argument against the literal signature of the wrapper.
+    run_call_with_unpacked_inputs.__signature__ = original_signature
+
+    return run_call_with_unpacked_inputs
+
+
 def input_processing(func, config, input_ids, **kwargs):
```
Comment (patrickvonplaten) [inline src/transformers/modeling_tf_utils.py:387]: Is this function still needed?

Comment (patrickvonplaten) [inline src/transformers/modeling_tf_utils.py:387]: Is this function still needed?

**Comment by patrickvonplaten** [comment]
Looks very nice! Thanks for cleaning this up

**Comment by patrickvonplaten** [comment]
Looks very nice! Thanks for cleaning this up

**Comment by gante** [comment]
It seems like everyone is in agreement :) @Rocketknight1 can you have a final look, and approve if you agree with the change?

**Review by Rocketknight1 — APPROVED** [review]
This looks great to me! Is the plan to wait and see how things go with BERT, and assuming no problems to repeat those edits across all model classes?

**Comment by gante** [comment]
> Is the plan to wait and see how things go with BERT, and assuming no problems to repeat those edits across all model classes?

Yeah! I'm also going to open an issue with the `good first issue` tag and try to get contributors in to replicate the change over other models. After a month or so, I will update any missing model.
